src/main/java/net/dv8tion/jda/bot/entities/impl/DefaultShardManagerImpl.java

```diff
@@ -0,0 +1,336 @@
+        {
+            final JDA jda = this.shards.getMap().remove(shardId);
+            if (jda != null)
+            {
+                jda.shutdown();
+                this.queue.offer(shardId);
+            }
+        }
+    }
+
+    @Override
+    public void restart(final int shardId)
+    {
+        Checks.notNegative(shardId, "shardId");
+        Checks.check(shardId < this.shardsTotal, "shardId must be lower than shardsTotal");
+
+        final JDA jda = this.shards.getMap().remove(shardId);
+        if (jda != null)
+            jda.shutdown();
+
+        this.queue.add(shardId);
+    }
+
+    @Override
+    public void shutdown()
+    {
+        if (this.shutdown.getAndSet(true))
+            return; // shutdown has already been requested
+
+        if (this.worker != null && !this.worker.isDone())
+            this.worker.cancel(true);
+
+        if (this.shutdownHook != null)
+        {
+            try
+            {
+                Runtime.getRuntime().removeShutdownHook(this.shutdownHook);
+            }
+            catch (final Exception ignored) {}
+        }
+
+        this.executor.shutdown();
+
+        if (this.shards != null)
+        {
+            for (final JDA jda : this.shards)
+                jda.shutdown();
+        }
+    }
+
+    @Override
+    public void shutdown(final int shardId)
+    {
+        final JDA jda = this.shards.getMap().remove(shardId);
+        if (jda != null)
+            jda.shutdown();
+    }
+
+    @Override
+    public void start(final int shardId)
+    {
+        Checks.notNegative(shardId, "shardId");
+        Checks.check(shardId < this.shardsTotal, "shardId must be lower than shardsTotal");
+        this.queue.add(shardId);
+    }
+
+    protected JDAImpl buildInstance(final int shardId) throws LoginException, RateLimitedException
+    {
+        final JDAImpl jda = new JDAImpl(AccountType.BOT, this.httpClientBuilder, this.wsFactory, this.shardedRateLimiter, this.autoReconnect,
+                this.enableVoice, false, this.enableBulkDeleteSplitting, this.corePoolSize, this.maxReconnectDelay);
+
+        jda.asBot().setShardManager(this);
+
+        if (this.eventManager != null)
+            jda.setEventManager(this.eventManager);
+
+        if (this.audioSendFactory != null)
+            jda.setAudioSendFactory(this.audioSendFactory);
+
+        this.listeners.forEach(jda::addEventListener);
+        jda.setStatus(JDA.Status.INITIALIZED); //This is already set by JDA internally, but this is to make sure the listeners catch it.
+
+        // Set the presence information before connecting to have the correct information ready when sending IDENTIFY
+        ((PresenceImpl) jda.getPresence()).setCacheGame(this.game).setCacheIdle(this.idle).setCacheStatus(this.status);
+
+        final JDAImpl.ShardInfoImpl shardInfo = new JDAImpl.ShardInfoImpl(shardId, this.shardsTotal);
+
+        final int shardTotal = jda.login(this.token, shardInfo, this.reconnectQueue);
```

Comment:
If I see this correctly this will use the GET_GATEWAY route on every shard? I think we can optimize this here.

Reply:
have to think about the best way to make this working with #520